src/Compilers/CSharp/Portable/Lowering/Instrumentation/LocalStateTracingInstrumenter.cs

```diff
@@ -0,0 +1,508 @@
+            BindingDiagnosticBag diagnostics,
+            Instrumenter previous)
+            : base(previous)
+        {
+            _scope = state;
+            _instrumentationType = instrumentationType;
+            _factory = factory;
+            _diagnostics = diagnostics;
+        }
+
+        protected override CompoundInstrumenter WithPreviousImpl(Instrumenter previous)
+            => new LocalStateTracingInstrumenter(
+                _scope,
+                _instrumentationType,
+                _factory,
+                _diagnostics,
+                previous);
+
+        public static bool TryCreate(
+            MethodSymbol method,
+            BoundStatement methodBody,
+            SyntheticBoundNodeFactory factory,
+            BindingDiagnosticBag diagnostics,
+            Instrumenter previous,
+            [NotNullWhen(true)] out LocalStateTracingInstrumenter? instrumenter)
+        {
+            instrumenter = null;
+
+            // Do not instrument implicitly-declared methods, except for constructors.
+            // Instrument implicit constructors in order to instrument member initializers.
+            if (method.IsImplicitlyDeclared && !method.IsImplicitConstructor)
+            {
+                return false;
+            }
+
+            // Method has no user-defined body
+            if (method is SourceMemberMethodSymbol { Bodies: { arrowBody: null, blockBody: null } } and not SynthesizedSimpleProgramEntryPointSymbol)
+            {
+                return false;
+            }
+
+            var instrumentationType = factory.Compilation.GetWellKnownType(WellKnownType.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker);
+            if (method.ContainingType.Equals(instrumentationType))
+            {
+                return false;
+            }
+
+            var scope = new Scope(factory.SynthesizedLocal(instrumentationType, methodBody.Syntax, kind: SynthesizedLocalKind.LocalStoreTracker));
+            instrumenter = new LocalStateTracingInstrumenter(scope, instrumentationType, factory, diagnostics, previous);
+            return true;
+        }
+
+        private MethodSymbol? GetLocalOrParameterStoreLogger(TypeSymbol variableType, Symbol targetSymbol, bool? refAssignmentSourceIsLocal, SyntaxNode syntax)
+        {
+            var enumDelta = (targetSymbol.Kind == SymbolKind.Parameter) ?
+                WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogParameterStoreBoolean - WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogLocalStoreBoolean : 0;
+
+            var overload = refAssignmentSourceIsLocal switch
+            {
+                true => WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogLocalStoreLocalAlias,
+                false => WellKnownMember.Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogLocalStoreParameterAlias,
+                null => variableType.EnumUnderlyingTypeOrSelf().SpecialType switch
```

Comment:
>switch [](http://example.com/codeflow?start=76&length=6)

Do we need to handle `System_IntPtr` and `System_UIntPtr` for `nint` and `nuint`?

Reply:
`nint` and `nuint` are handled by `Microsoft_CodeAnalysis_Runtime_LocalStoreTracker__LogLocalStoreUnmanaged`. See tests here:

https://github.com/dotnet/roslyn/pull/66768/files/f6ef35d90945bc07c5fa6e4a949c3d92e8eb4101#diff-ae134a952b809eb3609c40de1b6e727a013460652a73943cda7c3101e2b954c5R1067-R1068